### PR TITLE
Bug 1364391 – Add `UNNotificationServiceExtension`.

### PR DESCRIPTION
**Pull request by jhugman**
This PR introduces a `UNNotificationServiceExtension` to catch `mutable-content` APNS from the development push server.

This prepares the way for experimenting with the `aps` block. This should not be merged in its current form (i.e. the sending of an `aps` block in `PushClient.swift`, and the use `LatestDevFirefoxAccountConfiguration` instead of `StageFirefoxAccountConfiguration` in `Profile.swift` should not be merged). 

To test, the intent is that you should be able to switch to `LatestDevFirefoxAccountConfiguration` in Debug settings (navigate to Settings, close your eyes, tap five times on the Build version, then toggle the "Use Staging Sync Service"), then create an account on that server, then log in. This will only work with the `Fennec` scheme.

For other schemes, switching to "Use Staging Sync Service" will use the staging firefox account configuration.

Currently, this is the only combination of 1. Scheme, 2. FxA Configuration 3. autopush which is mututally compatible.

After a deploy cycle of FxA and autopush (and finding the correct `aps` block), this should be possible with production builds.

Currently, the `aps` block is only stored at registration by the `FennecPushConfiguration`. Thus, until the Push team change the default `aps` block on other instances, this testing will only work with the `Fennec` scheme.

https://bugzilla.mozilla.org/show_bug.cgi?id=1364391

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
Profile.swift | 433 | Force casts should be avoided. 
Profile.swift | 84 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by jhugman** [comment]
Tied toggling of account & push config back to `Fennec` scheme as suggested.

@farhanpatel Please could you re-review?

File: Extensions/NotificationService/NotificationService.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    var contentHandler: ((UNNotificationContent) -> Void)?
+    var bestAttemptContent: UNMutableNotificationContent?
+
+    override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
+        self.contentHandler = contentHandler
+        bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
+
+        var userInfo = request.content.userInfo
+        userInfo["url"] = "https://mozilla.org"
+        NSLog("NotificationService APNS NOTIFICATION \(userInfo)")
```
Comment (farhanpatel) [inline Extensions/NotificationService/NotificationService.swift]: Is there a reason why we have these NSLogs? We could probably use our regular logs.

File: Extensions/NotificationService/NotificationService.swift [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    var contentHandler: ((UNNotificationContent) -> Void)?
+    var bestAttemptContent: UNMutableNotificationContent?
+
+    override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
+        self.contentHandler = contentHandler
+        bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
+
+        var userInfo = request.content.userInfo
+        userInfo["url"] = "https://mozilla.org"
```
Comment (farhanpatel) [inline Extensions/NotificationService/NotificationService.swift]: Is this suppose to be hard coded?

Comment (jhugman) [inline Extensions/NotificationService/NotificationService.swift]: This is supposed to be hardcoded. There's a lot of work to be done yet.

File: Extensions/NotificationService/NotificationService.swift (hunk 1 above)
Comment (jhugman) [inline Extensions/NotificationService/NotificationService.swift]: NSLog appears in the Console.app. 

We can't see out logs when logging from the extension. (we can, but we want to be able to see both app and extension at the same time).

This is temporary code. It is currently tightly controlled so as not to be accessible from release.

**Pull request merged**